# Data Description 


Estimating churners prior to their abandonment of a product or service is crucial. In this machine learning assignment, I will design a churn prediction model for the telecom industry to forecast which customers are most likely to churn.

The 39 features of this dataset are as follows:

1. Acronyms: Descriptions
2. MOBILE_NUMBER: Customer phone number
3. CIRCLE_ID: Telecom circle area to which the customer belongs to
4. LOC	Local calls: within same telecom circle
5. STD	STD calls: outside the calling circle
6. IC: Incoming calls
7. OG: Outgoing calls
8. T2T: Operator T to T, i.e. within same operator (mobile to mobile)
9. T2M: Operator T to other operator mobile
10. T2O: Operator T to other operator fixed line
11. T2F: Operator T to fixed lines of T
12. T2C: Operator T to it’s own call center
13. ARPU: Average revenue per user
14. MOU: Minutes of usage - voice calls
15. AON: Age on network - number of days the customer is using the operator T network
16. ONNET: All kind of calls within the same operator network
17. OFFNET:    	All kind of calls outside the operator T network
18. ROAM:	Indicates that customer is in roaming zone during the call
19. SPL:   	Special calls
20. ISD:    	ISD calls
21. RECH:    	Recharge
22. NUM:    	Number
23. AMT:    	Amount in local currency
24. MAX:    	Maximum
25. DATA:    	Mobile internet
26. 3G:    	3G network
27. AV:    	Average
28. VOL:    	Mobile internet usage volume (in MB)
29. 2G:    	2G network
30. PCK:    	Prepaid service schemes called - PACKS
31. NIGHT:    	Scheme to use during specific night hours only
32. MONTHLY:    	Service schemes with validity equivalent to a month
33. SACHET:   	Service schemes with validity smaller than a month
34. *.6:    	KPI for the month of June
35. *.7:    	KPI for the month of July
36. *.8:    	KPI for the month of August
37. *.9:    	KPI for the month of September
38. FB_USER:	Service scheme to avail services of Facebook and similar social networking sites
39. VBC:    	Volume based cost - when no specific scheme is not purchased and paid as per usage

# Problem Statement

In the telecommunications business, clients may pick from a variety of service providers and actively switch from one to another. In this extremely competitive sector, the yearly turnover rate for the telecoms industry averages between 15 and 25 percent. Given that it costs five to ten times as much to gain a new client as it does to maintain an existing one, customer retention is now more crucial than customer acquisition.

Retaining high-profitable clients is the primary business objective for many incumbent operators of T. To decrease customer turnover, telecom businesses must anticipate which consumers are at high risk of churning.

# Aim

In this project, I am going to analyze the customer-level data of a large telecommunications company, develop predictive models to identify consumers at high risk of churn, and identify the primary signs of churn.


# Tech stack

* Language – Python
* Libraries - Numpy, Pandas, Matplotlib, Seaborn, Scikit-learn,

# Approach

1. Importing the required libraries and reading the dataset.
    * Understanding the dataset
2. Exploratory Data Analysis (EDA) –
3. Filtering High Value Customers
4. Creating target Variable
5. Developing new features
6. Handling Missing values
7. Data Visualization-Univariate Analysis
8. Data Visualization- Bivariate Analysis
9. Outlier Detection
10. Data Preparation
11. Data Modeling and Eavlaution
12. Non-Interpretable Models
13. Interpretable Models
14. Conclusion

# 1. Understanding the dataset

In [1]:
%%HTML
<style type="text/css">
table.dataframe td, table.dataframe th {
    border: 1px  black solid !important;
  color: black !important;
}
</style>

## 1.1 Importing Libraries

In [2]:
!pip install graphviz

In [3]:
#Importing Data Reading and Processing Libraries
import pandas as pd
import numpy as np

#Imporitng Data Visualization Libraries
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

#Importing Data Preparation and Modeling Libraries
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold, GridSearchCV,StratifiedKFold

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier


#Importing Warning Libraries
import warnings
warnings.filterwarnings("ignore")

#Importing Miscellaneous Libraries
pd.set_option("display.max_columns",None)
pd.set_option("display.max_rows",None)
pd.set_option('display.width', None)

# Import the StandardScaler()
from sklearn.preprocessing import StandardScaler

#Improting the PCA module
from sklearn.decomposition import PCA

# For Hopkins test
from sklearn.neighbors import NearestNeighbors
from random import sample
from numpy.random import uniform
import numpy as np
from math import isnan

# For clustering 
## using KMeans ##
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

# Importing classification report and confusion matrix from sklearn metrics
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.metrics import recall_score,roc_auc_score,roc_curve

## using Hierarchical ##
from scipy.cluster.hierarchy import linkage
from scipy.cluster.hierarchy import dendrogram
from scipy.cluster.hierarchy import cut_tree

# Importing required packages for visualization
from IPython.display import Image  
#from sklearn.externals.six import StringIO  
from sklearn.tree import export_graphviz
import pydot, graphviz


# Other sklearn packages
import sklearn.metrics as metrics
from xgboost import XGBClassifier
from sklearn.tree import DecisionTreeClassifier

from datetime import date,datetime
import math
import multiprocessing

print("Successfully Importing Libraries...")

Successfully Importing Libraries...


# 1.2 Dataset Loading 

In [4]:
df = pd.read_csv('telecom_churn_data.csv')
print("Successully Datset Load..")
df.sample(2)

Successully Datset Load..


mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
9574      7000122759        109             0.0             0.0   
26343     7000248369        109             0.0             0.0   

       loc_ic_t2o_mou last_date_of_month_6 last_date_of_month_7  \
9574              0.0            6/30/2014            7/31/2014   
26343             0.0            6/30/2014            7/31/2014   

      last_date_of_month_8 last_date_of_month_9   arpu_6   arpu_7   arpu_8  \
9574             8/31/2014            9/30/2014  472.709  429.116  326.263   
26343            8/31/2014            9/30/2014  263.280  154.144   88.000   

        arpu_9  onnet_mou_6  onnet_mou_7  onnet_mou_8  onnet_mou_9  \
9574   351.012        82.66        19.93        23.21        23.41   
26343   86.000         9.19        33.46        65.36        47.81   

       offnet_mou_6  offnet_mou_7  offnet_mou_8  offnet_mou_9  roam_ic_mou_6  \
9574         167.13         90.21         65.13         58.14            0.0   
26343          0.90          5.49          1.40          3.98            0.0   

       roam_ic_mou_7  roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  \
9574             0.0            0.0            0.0            0.0   
26343            0.0            0.0            0.0            0.0   

       roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  loc_og_t2t_mou_6  \
9574             0.0            0.0            0.0             29.44   
26343            0.0            0.0            0.0              9.19   

       loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  loc_og_t2m_mou_6  \
9574              12.14              9.21             12.64            116.33   
26343             33.46             65.36             47.81              0.90   

       loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  loc_og_t2f_mou_6  \
9574              89.31             59.94             54.66               0.0   
26343              5.49              1.40              3.98               0.0   

       loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  loc_og_t2c_mou_6  \
9574                0.0              5.18              0.46               0.0   
26343               0.0              0.00              0.00               0.0   

       loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  loc_og_mou_6  \
9574                0.0               0.0              2.26        145.78   
26343               0.0               0.0              0.00         10.09   

       loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
9574         101.46         74.34         67.78             53.21   
26343         38.96         66.76         51.79              0.00   

       std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2t_mou_9  std_og_t2m_mou_6  \
9574               7.78             13.99             10.76             47.11   
26343              0.00              0.00              0.00              0.00   

       std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2m_mou_9  std_og_t2f_mou_6  \
9574                0.9               0.0              0.75               0.0   
26343               0.0               0.0              0.00               0.0   

       std_og_t2f_mou_7  std_og_t2f_mou_8  std_og_t2f_mou_9  std_og_t2c_mou_6  \
9574                0.0               0.0               0.0               0.0   
26343               0.0               0.0               0.0               0.0   

       std_og_t2c_mou_7  std_og_t2c_mou_8  std_og_t2c_mou_9  std_og_mou_6  \
9574                0.0               0.0               0.0        100.33   
26343               0.0               0.0               0.0          0.00   

       std_og_mou_7  std_og_mou_8  std_og_mou_9  isd_og_mou_6  isd_og_mou_7  \
9574           8.68         13.99         11.51           0.0           0.0   
26343          0.00          0.00          0.00           0.0           0.0   

       isd_og_mou_8  isd_og_mou_9  spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  \
9574            0.0           0.0     

In [5]:
df.shape

(99999, 226)

`Observation`: Number of rows: 99999, 
               Number of columns: 226 

In [6]:
df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 226 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    mobile_number             int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    last_date_of_month_9      object 
 9    arpu_6                    float64
 10   arpu_7                    float64
 11   arpu_8                    float64
 12   arpu_9                    float64
 13   onnet_mou_6               float64
 14   onnet_mou_7               float64
 15   onnet_mou_8               float64
 16   onnet_mou_9               float64
 17   offnet_mou_6              float64
 18   offnet_mou_7              float64
 19   offnet_mou_8              float64
 20   offn

In [7]:
#Let's examine the data's distribution.
df.describe().T

count          mean            std           min  \
mobile_number       99999.0  7.001207e+09  695669.386290  7.000000e+09   
circle_id           99999.0  1.090000e+02       0.000000  1.090000e+02   
loc_og_t2o_mou      98981.0  0.000000e+00       0.000000  0.000000e+00   
std_og_t2o_mou      98981.0  0.000000e+00       0.000000  0.000000e+00   
loc_ic_t2o_mou      98981.0  0.000000e+00       0.000000  0.000000e+00   
arpu_6              99999.0  2.829874e+02     328.439770 -2.258709e+03   
arpu_7              99999.0  2.785366e+02     338.156291 -2.014045e+03   
arpu_8              99999.0  2.791547e+02     344.474791 -9.458080e+02   
arpu_9              99999.0  2.616451e+02     341.998630 -1.899505e+03   
onnet_mou_6         96062.0  1.323959e+02     297.207406  0.000000e+00   
onnet_mou_7         96140.0  1.336708e+02     308.794148  0.000000e+00   
onnet_mou_8         94621.0  1.330181e+02     308.951589  0.000000e+00   
onnet_mou_9         92254.0  1.303023e+02     308.477668  0.000000e+00   
offnet_mou_6        96062.0  1.979356e+02     316.851613  0.000000e+00   
offnet_mou_7        96140.0  1.970451e+02     325.862803  0.000000e+00   
offnet_mou_8        94621.0  1.965748e+02     327.170662  0.000000e+00   
offnet_mou_9        92254.0  1.903372e+02     319.396092  0.000000e+00   
roam_ic_mou_6       96062.0  9.950013e+00      72.825411  0.000000e+00   
roam_ic_mou_7       96140.0  7.149898e+00      73.447948  0.000000e+00   
roam_ic_mou_8       94621.0  7.292981e+00      68.402466  0.000000e+00   
roam_ic_mou_9       92254.0  6.343841e+00      57.137537  0.000000e+00   
roam_og_mou_6       96062.0  1.391134e+01      71.443196  0.000000e+00   
roam_og_mou_7       96140.0  9.818732e+00      58.455762  0.000000e+00   
roam_og_mou_8       94621.0  9.971890e+00      64.713221  0.000000e+00   
roam_og_mou_9       92254.0  8.555519e+00      58.438186  0.000000e+00   
loc_og_t2t_mou_6    96062.0  4.710076e+01     150.856393  0.000000e+00   
loc_og_t2t_mou_7    96140.0  4.647301e+01     155.318705  0.000000e+00   
loc_og_t2t_mou_8    94621.0  4.588781e+01     151.184830  0.000000e+00   
loc_og_t2t_mou_9    92254.0  4.458445e+01     147.995390  0.000000e+00   
loc_og_t2m_mou_6    96062.0  9.334209e+01     162.780544  0.000000e+00   
loc_og_t2m_mou_7    96140.0  9.139713e+01     157.492308  0.000000e+00   
loc_og_t2m_mou_8    94621.0  9.175513e+01     156.537048  0.000000e+00   
loc_og_t2m_mou_9    92254.0  9.046319e+01     158.681454  0.000000e+00   
loc_og_t2f_mou_6    96062.0  3.751013e+00      14.230438  0.000000e+00   
loc_og_t2f_mou_7    96140.0  3.792985e+00      14.264986  0.000000e+00   
loc_og_t2f_mou_8    94621.0  3.677991e+00      13.270996  0.000000e+00   
loc_og_t2f_mou_9    92254.0  3.655123e+00      13.457549  0.000000e+00   
loc_og_t2c_mou_6    96062.0  1.123056e+00       5.448946  0.000000e+00   
loc_og_t2c_mou_7    96140.0  1.368500e+00       7.533445  0.000000e+00   
loc_og_t2c_mou_8    94621.0  1.433821e+00       6.783335  0.000000e+00   
loc_og_t2c_mou_9    92254.0  1.232726e+00       5.619021  0.000000e+00   
loc_og_mou_6        96062.0  1.442012e+02     251.751489  0.000000e+00   
loc_og_mou_7        96140.0  1.416705e+02     248.731086  0.000000e+00   
loc_og_mou_8        94621.0  1.413282e+02     245.914311  0.000000e+00   
loc_og_mou_9        92254.0  1.387100e+02     245.934517  0.000000e+00   
std_og_t2t_mou_6    96062.0  7.982987e+01     252.476533  0.000000e+00   
std_og_t2t_mou_7    96140.0  8.329960e+01     263.631042  0.000000e+00   
std_og_t2t_mou_8    94621.0  8.328267e+01     265.486090  0.000000e+00   
std_og_t2t_mou_9    92254.0  8.234292e+01     267.184991  0.000000e+00   
std_og_t2m_mou_6    96062.0  8.729962e+01     255.617850  0.000000e+00   
std_og_t2m_mou_7    96140.0  9.080414e+01     269.347911  0.000000e+00   
std_og_t2m_mou_8    94621.0  8.983839e+01     271.757783  0.000000e+00   
std_og_t2m_mou_9    92254.0  8.627662e+01     261.407396  0.000000e+00   
std_og_t2f_mou_6    96062.

# 2. Data Cleaning

Some columns representing volume-based users include the month in their names, whereas the remaining columns use the numerals "6,7,8" to indicate the month. I am planning to rename the missplead column.

In [8]:
month = ['aug_vbc_3g','jul_vbc_3g','jun_vbc_3g','sep_vbc_3g']
df = df.rename(columns = {'aug_vbc_3g':'vbc_3g_8','jul_vbc_3g':'vbc_3g_7','jun_vbc_3g':'vbc_3g_6',
                          'sep_vbc_3g':'vbc_3g_9'})


## 2.1 Converting Datetime

In [9]:
#The conversion of date columns to date-time format

date_col= [col for col in df.columns if 'date' in col]

for i in df[date_col]:
    df[i] = pd.to_datetime(df[i])


In [10]:
df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 226 columns):
 #    Column                    Dtype         
---   ------                    -----         
 0    mobile_number             int64         
 1    circle_id                 int64         
 2    loc_og_t2o_mou            float64       
 3    std_og_t2o_mou            float64       
 4    loc_ic_t2o_mou            float64       
 5    last_date_of_month_6      datetime64[ns]
 6    last_date_of_month_7      datetime64[ns]
 7    last_date_of_month_8      datetime64[ns]
 8    last_date_of_month_9      datetime64[ns]
 9    arpu_6                    float64       
 10   arpu_7                    float64       
 11   arpu_8                    float64       
 12   arpu_9                    float64       
 13   onnet_mou_6               float64       
 14   onnet_mou_7               float64       
 15   onnet_mou_8               float64       
 16   onnet_mou_9               float64     

# 2.2 Removing single unique value
Columns with a single unique value for all clients are essentially meaningless. As a result, i am deleting columns with zero variance from the dataset.

In [11]:
cols = []
for i in df.columns:
    if df[i].nunique() ==1:
        cols.append(i)
df = df.drop(cols,axis=1)
df.sample(2)

mobile_number   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
25886     7001898209  223.495  244.635  142.680  229.555         0.00   
8724      7001078801  346.067  317.799  229.246  144.230        63.81   

       onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
25886         0.00         0.00        13.14         29.74         52.81   
8724         27.58        53.31        20.94        317.78        342.66   

       offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  \
25886         10.33         13.41           0.00           0.00   
8724         228.36        134.66          10.69           6.28   

       roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  \
25886            0.0            0.0           0.00           0.00   
8724             0.0            0.0         153.88          98.33   

       roam_og_mou_8  roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
25886            0.0            0.0              0.00              0.00   
8724             0.0            0.0             32.61             14.58   

       loc_og_t2t_mou_8  loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
25886              0.00             13.14              7.23             49.43   
8724              36.18             14.43            134.81            147.91   

       loc_og_t2m_mou_8  loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
25886              8.68             13.41              1.56              1.46   
8724             120.44            111.28              0.00              0.00   

       loc_og_t2f_mou_8  loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
25886              1.65               0.0               0.0               0.0   
8724               3.74               0.0               0.0               0.0   

       loc_og_t2c_mou_8  loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  \
25886               0.0               0.0          8.79         50.89   
8724                0.0               0.0        167.43        162.49   

       loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  std_og_t2t_mou_7  \
25886         10.33         26.56              0.00              0.00   
8724         160.38        125.71             17.76              0.16   

       std_og_t2t_mou_8  std_og_t2t_mou_9  std_og_t2m_mou_6  std_og_t2m_mou_7  \
25886              0.00              0.00              0.00              0.00   
8724              17.13              6.51             42.51            109.24   

       std_og_t2m_mou_8  std_og_t2m_mou_9  std_og_t2f_mou_6  std_og_t2f_mou_7  \
25886              0.00              0.00               0.0               0.0   
8724             104.16             23.38               0.0               0.0   

       std_og_t2f_mou_8  std_og_t2f_mou_9  std_og_mou_6  std_og_mou_7  \
25886               0.0               0.0          0.00          0.00   
8724                0.0               0.0         60.28        109.41   

       std_og_mou_8  std_og_mou_9  isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  \
25886          0.00          0.00           0.0           0.0           0.0   
8724         121.29         29.89           0.0           0.0           0.0   

       isd_og_mou_9  spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  spl_og_mou_9  \
25886           0.0         20.95          1.91           0.0           0.0   
8724            0.0          0.00          0.00           0.7           0.0   

       og_others_6  og_others_7  og_others_8  og_others_9  total_og_mou_6  \
25886          0.0          0.0          0.0          0.0           29.74   
8724           0.0          0.0          0.0          0.0          227.71   

       total_og_mou_7  total_og_mou_8  total_og_mou_9  loc_ic_t2t_mou_6  \
25886           52.81           10.33           26.56             32.34   
8724           271.91          282.38          155.61            158.61   

       loc_ic_t2t_mou_7  loc_ic_t2t_mou_8  loc_ic_t2t_mou_9  loc_ic_t2m_mou_6  \
25886              7.39    

In [12]:
df.shape

(99999, 210)

`Observation`: Number of columns: unique column 16 is dropped from 226 to 210. 

# 2.3 Create ID for each customer

Eliminate the mobile number column and create an id column for client identification.

In [13]:
#drop mobile number
df = df.drop('mobile_number', axis=1)

#Create id
df = df.reset_index()
df = df.rename(columns = {'index':'cust_id'})
df['cust_id'] = df['cust_id']+1
df.sample(2)

cust_id   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  onnet_mou_7  \
37801    37802  311.469  239.701  202.850  216.139        93.03         0.03   
98814    98815  180.040  505.767   86.507  255.502         9.13         0.00   

       onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
37801        38.46        36.83        598.99        472.61        329.43   
98814         0.00         7.26        171.44         53.89         24.78   

       offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
37801        286.54           0.00            0.0            0.0   
98814         75.46          46.26            0.0            0.0   

       roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
37801            0.0           0.00            0.0            0.0   
98814            0.0          61.26            0.0            0.0   

       roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
37801            0.0              1.54              0.03               0.0   
98814            0.0              0.00              0.00               0.0   

       loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
37801              0.00             33.14             16.18             16.23   
98814              7.26             63.98             45.66             24.78   

       loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
37801             15.21              0.38              0.00               0.0   
98814             74.71              7.34              6.19               0.0   

       loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
37801              0.00              0.00              0.00             11.68   
98814              0.75              5.04              1.99              0.00   

       loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
37801               0.0         35.08         16.21         16.23   
98814               0.0         71.33         51.86         24.78   

       loc_og_mou_9  std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  \
37801         15.21             91.48               0.0             38.46   
98814         82.73              8.95               0.0              0.00   

       std_og_t2t_mou_9  std_og_t2m_mou_6  std_og_t2m_mou_7  std_og_t2m_mou_8  \
37801             36.83            565.46            456.43            284.39   
98814              0.00             29.06              0.03              0.00   

       std_og_t2m_mou_9  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  \
37801            270.96               0.0               0.0               0.0   
98814              0.00               0.0               0.0               0.0   

       std_og_t2f_mou_9  std_og_mou_6  std_og_mou_7  std_og_mou_8  \
37801               0.0        656.94        456.43        322.86   
98814               0.0         38.01          0.03          0.00   

       std_og_mou_9  isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  isd_og_mou_9  \
37801        307.79           0.0           0.0          0.03           0.0   
98814          0.00           0.0           0.0          0.00           0.0   

       spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  spl_og_mou_9  og_others_6  \
37801          0.00          0.00         28.76          0.36          0.0   
98814          9.96          5.73          0.60          0.76          0.0   

       og_others_7  og_others_8  og_others_9  total_og_mou_6  total_og_mou_7  \
37801          0.0          0.0          0.0          692.03          472.64   
98814          0.0          0.0          0.0          119.31           57.63   

       total_og_mou_8  total_og_mou_9  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  \
37801          367.89          323.38              1.16              0.38   
98814           25.38           83.49              0.00              0.00   

       loc_ic_t2t_mou_8  loc_ic_t2t_mou_9  loc_ic_t2m_mou_6  loc_ic_t2m_

## 2.4 Separate Categorical columns

Check if this dataset has any category columns. This may be determined by identifying which columns contain the numbers 1 and 0. These values are equivalent to yes and no, respectively.

In [14]:
cat_cols = []

for i in df.columns:
    if df[i].nunique()==2:
        cat_cols.append(i)
cat_cols 

['night_pck_user_6',
 'night_pck_user_7',
 'night_pck_user_8',
 'night_pck_user_9',
 'fb_user_6',
 'fb_user_7',
 'fb_user_8',
 'fb_user_9']

## 2.5 Some Important Insights

### 2.5.1 Avaerage Revenue of call per user

In [15]:
arpu=[]
arpu_col= [col for col in df.columns if 'arpu' in col]

for i in df[arpu_col]:
    x=df[[i]].min()
    arpu.append(x)
print(arpu)

[arpu_6   -2258.709
dtype: float64, arpu_7   -2014.045
dtype: float64, arpu_8   -945.808
dtype: float64, arpu_9   -1899.505
dtype: float64, arpu_3g_6   -30.82
dtype: float64, arpu_3g_7   -26.04
dtype: float64, arpu_3g_8   -24.49
dtype: float64, arpu_3g_9   -71.09
dtype: float64, arpu_2g_6   -35.83
dtype: float64, arpu_2g_7   -15.48
dtype: float64, arpu_2g_8   -55.83
dtype: float64, arpu_2g_9   -45.74
dtype: float64]


`Observation`
The minimum values in arpu 6, arpu 7, arpu 8, and arpu 9 are seen to be negative. This suggests that certain clients are causing a loss for the business. I will maintain them in my study since my research criteria for a high-value client are based on usage-based churn, not revenue-based churn. Getting rid of them might result in the loss of some insightful knowledge. Observe their significance in the exploratory data analysis section before making a decision.

### 2.5.2 Sachet recharge
Recharge sachets are Service plans having a duration of less than one month. This indicates that the number of days needed to recharge a sachet should be fewer than 30. Any service plans that extend beyond 29 days should indicate that the client has performed a monthly recharge or that the entry is incorrect. Let us limit the values beyond 29 days to the maximum number of days recharged below 30 days.

In [16]:
sachet=[]
sachet_col= [col for col in df.columns if 'sachet' in col]

for i in df[sachet_col]:
    x=df[[i]].max()
    sachet.append(x)
print(sachet)

[sachet_2g_6    42
dtype: int64, sachet_2g_7    48
dtype: int64, sachet_2g_8    44
dtype: int64, sachet_2g_9    40
dtype: int64, sachet_3g_6    29
dtype: int64, sachet_3g_7    35
dtype: int64, sachet_3g_8    41
dtype: int64, sachet_3g_9    49
dtype: int64]


In [17]:
df['sachet_2g_6'] = df['sachet_2g_6'].clip(0,28)
df['sachet_2g_7'] = df['sachet_2g_7'].clip(0,29)
df['sachet_2g_8'] = df['sachet_2g_8'].clip(0,29)
df['sachet_3g_6'] = df['sachet_3g_6'].clip(0,29)
df['sachet_3g_7'] = df['sachet_3g_7'].clip(0,24)
df['sachet_3g_8'] = df['sachet_3g_8'].clip(0,29)

In [18]:
df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 210 columns):
 #    Column                    Dtype         
---   ------                    -----         
 0    cust_id                   int64         
 1    arpu_6                    float64       
 2    arpu_7                    float64       
 3    arpu_8                    float64       
 4    arpu_9                    float64       
 5    onnet_mou_6               float64       
 6    onnet_mou_7               float64       
 7    onnet_mou_8               float64       
 8    onnet_mou_9               float64       
 9    offnet_mou_6              float64       
 10   offnet_mou_7              float64       
 11   offnet_mou_8              float64       
 12   offnet_mou_9              float64       
 13   roam_ic_mou_6             float64       
 14   roam_ic_mou_7             float64       
 15   roam_ic_mou_8             float64       
 16   roam_ic_mou_9             float64     

# 3 Focusing on high value Customers

In [19]:
df['total_amt_6'] = df[['total_rech_amt_6', 'total_rech_data_6']].sum(axis=1)
df['total_amt_7'] = df[['total_rech_amt_7', 'total_rech_data_7']].sum(axis=1)
df['total_amt_8'] = df[['total_rech_amt_8', 'total_rech_data_8']].sum(axis=1)
df['total_amt_9'] = df[['total_rech_amt_9', 'total_rech_data_9']].sum(axis=1)

In [20]:
#calculate total recharge amount per user for months
df['total_rech_amt_per_user']=df[['total_amt_6','total_amt_7','total_amt_8','total_amt_9']].sum(axis=1)

In [21]:
df.sample(2)

cust_id   arpu_6   arpu_7   arpu_8  arpu_9  onnet_mou_6  onnet_mou_7  \
89817    89818   62.307   37.770   44.769  44.318         9.71        10.94   
44657    44658  537.145  476.488  210.044   0.000       138.99       209.99   

       onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
89817        10.06        18.24         32.13         25.83         30.61   
44657        57.66          NaN       1082.81        883.58        159.08   

       offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
89817         19.51            0.0            0.0            0.0   
44657           NaN            0.0            0.0            0.0   

       roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
89817            0.0            0.0            0.0            0.0   
44657            NaN            0.0            0.0            0.0   

       roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
89817            0.0              5.78              0.56              0.53   
44657            NaN             49.68             48.24             36.71   

       loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
89817               0.0             19.11             20.99             27.38   
44657               NaN            304.78            265.39             41.24   

       loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
89817              9.78               0.0               0.0               0.0   
44657               NaN               0.0               0.0               0.0   

       loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
89817               0.0               0.0              4.83              2.48   
44657               NaN               0.0              0.00              0.00   

       loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
89817              1.56         24.89         21.56         27.91   
44657               NaN        354.46        313.64         77.96   

       loc_og_mou_9  std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  \
89817          9.78              3.93             10.38              9.53   
44657           NaN             89.31            161.74             20.94   

       std_og_t2t_mou_9  std_og_t2m_mou_6  std_og_t2m_mou_7  std_og_t2m_mou_8  \
89817             18.24             13.01              0.00              0.75   
44657               NaN            778.03            618.18            117.83   

       std_og_t2m_mou_9  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  \
89817              8.16               0.0               0.0               0.0   
44657               NaN               0.0               0.0               0.0   

       std_og_t2f_mou_9  std_og_mou_6  std_og_mou_7  std_og_mou_8  \
89817               0.0         16.94         10.38         10.28   
44657               NaN        867.34        779.93        138.78   

       std_og_mou_9  isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  isd_og_mou_9  \
89817         26.41           0.0           0.0           0.0           0.0   
44657           NaN           0.0           0.0           0.0           NaN   

       spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  spl_og_mou_9  og_others_6  \
89817          3.73          5.26          2.48          2.53         2.54   
44657          0.00          0.00          0.00           NaN         0.00   

       og_others_7  og_others_8  og_others_9  total_og_mou_6  total_og_mou_7  \
89817          0.0          0.0          0.0           48.13           37.21   
44657          0.0          0.0          NaN         1221.81         1093.58   

       total_og_mou_8  total_og_mou_9  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  \
89817           40.68           38.73             26.11              4.18   
44657          216.74            0.00             38.41             59.33   

       loc_ic_t2t_mou_8  loc_ic_t2t_mou_9  loc_ic_t2m_mou_6  loc_ic_t2m_mou

## 3.1 Filtering high value customers

In [22]:
avg_amt_rech=df['total_rech_amt_per_user'].mean()

In [23]:
high_val_customer = df.loc[df['total_rech_amt_per_user'] >= avg_amt_rech]
high_val_customer.head()

cust_id    arpu_6    arpu_7    arpu_8   arpu_9  onnet_mou_6  onnet_mou_7  \
3         4   221.338   251.102   508.054  389.500        99.91        54.39   
7         8  1069.180  1349.850  3171.480  500.000        57.84        54.68   
8         9   378.721   492.223   137.362  166.787       413.69       351.03   
13       14   492.846   205.671   593.260  322.732       501.76       108.39   
15       16    31.000   510.465   590.643  510.390          NaN       246.56   

    onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
3        310.98       241.71        123.31        109.01         71.68   
7         52.29          NaN        453.43        567.16        325.91   
8         35.08        33.46         94.66         80.63        136.48   
13       534.24       244.81        413.31        119.28        482.46   
15       280.31       289.79           NaN        839.58       1011.91   

    offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_ic_mou_9  \
3         113.54           0.00          54.86          44.38           0.00   
7            NaN          16.23          33.49          31.64            NaN   
8         108.71           0.00           0.00           0.00           0.00   
13        214.06          23.53         144.24          72.11         136.78   
15        642.14            NaN           0.00           0.00           0.00   

    roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
3            0.00          28.09          39.04           0.00   
7           23.74          12.59          38.06            NaN   
8            0.00           0.00           0.00           0.00   
13           7.98          35.26           1.44          12.78   
15            NaN           0.88           0.00           0.00   

    loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
3              73.68             34.81             10.61             15.49   
7              51.39             31.38             40.28               NaN   
8             297.13            217.59             12.49             26.13   
13             49.63              6.19             36.01              6.14   
15               NaN             16.13             44.79             48.33   

    loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
3             107.43             83.21             22.46             65.46   
7             308.63            447.38            162.28               NaN   
8              80.96             70.58             50.54             34.58   
13            151.13             47.28            294.46            108.24   
15               NaN             38.99             92.53            158.11   

    loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
3               1.91              0.65              4.91              2.06   
7              62.13             55.14             53.23               NaN   
8               0.00              0.00              0.00              0.00   
13              4.54              0.00             23.51              5.29   
15               NaN              5.13              4.83              8.43   

    loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  \
3                0.0              0.00              0.00              0.00   
7                0.0              0.00              0.00               NaN   
8                0.0              0.00              7.15              0.00   
13               0.0              0.00              0.49              0.00   
15               NaN              9.78              0.00              0.03   

    loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
3         183.03        118.68         37.99         83.03             26.23   
7         422.16        533.91        255.79           NaN              4.30   
8         378.09        288.18         63.04         60.71            116.56   
13  

In [24]:
print("Total number of high value customers: ",len(high_val_customer))

Total number of high value customers:  34958


# 4 Creating Target Variable¶

In this step, I will tag the churned customers whose target variable is categorised as churn Yes or churn No (Yes=1, No=0) based on T-four-month Mobile's data.


## 4.1 Find put inactive customers

During the churn period, those who have not made any calls (incoming or outgoing) and have not utilized mobile internet even once are considered to be inactive. The following attributes must be used to identify churners:

* total_ic_mou_6
* total_ic_mou_7
* total_ic_mou_8
* total_ic_mou_9
* total_og_mou_6
* total_og_mou_7
* total_og_mou_8
* total_og_mou_9
* vol_2g_mb_6
* vol_2g_mb_7
* vol_2g_mb_8
* vol_2g_mb_9
* vol_3g_mb_6
* vol_3g_mb_7
* vol_3g_mb_8
* vol_3g_mb_9

In [25]:
# The month of September is used to screen out churned customers.

df['churn'] = df.apply(lambda x: 1 if (x.total_ic_mou_9 == 0 and x.total_og_mou_9 == 0 and x.vol_2g_mb_9 ==0 and x.vol_3g_mb_9==0) else 0, axis=1)
df['churn'] = df['churn'].astype("str")
df.shape

(99999, 216)

In [26]:
# number of churned customers
df['churn'].value_counts()

0    89808
1    10191
Name: churn, dtype: int64

In [27]:
#what's the % of churned customers
print("The Percentage of churned customers is:" , round(100*(df.churn.astype("int").sum()/len(df)),2))

The Percentage of churned customers is: 10.19


### 4.2 After labeling churners, let's delete any characteristics pertaining to the churn phase (those with' 9', etc. in their names).

In [28]:
col_9 = [i for i in df.columns if '9' in i]
df = df.drop(col_9,axis=1)
df.shape

(99999, 163)

In [29]:
# let's update our categorical column list
cat_cols = [ele for ele in cat_cols if ele not in col_9]
cat_cols

['night_pck_user_6',
 'night_pck_user_7',
 'night_pck_user_8',
 'fb_user_6',
 'fb_user_7',
 'fb_user_8']

In [30]:
df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 163 columns):
 #    Column                    Dtype         
---   ------                    -----         
 0    cust_id                   int64         
 1    arpu_6                    float64       
 2    arpu_7                    float64       
 3    arpu_8                    float64       
 4    onnet_mou_6               float64       
 5    onnet_mou_7               float64       
 6    onnet_mou_8               float64       
 7    offnet_mou_6              float64       
 8    offnet_mou_7              float64       
 9    offnet_mou_8              float64       
 10   roam_ic_mou_6             float64       
 11   roam_ic_mou_7             float64       
 12   roam_ic_mou_8             float64       
 13   roam_og_mou_6             float64       
 14   roam_og_mou_7             float64       
 15   roam_og_mou_8             float64       
 16   loc_og_t2t_mou_6          float64     

In [31]:
df.sample(2)

cust_id   arpu_6   arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  \
50018    50019    2.500   27.368  172.104         0.00        36.74   
52686    52687  921.935  702.293  641.774       341.11        68.41   

       onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
50018       431.79          4.03          1.64         33.51            0.1   
52686       102.89        518.74        552.58        494.26            0.0   

       roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  \
50018           6.79           0.00           4.03          38.39   
52686           0.00          10.04           0.00           0.00   

       roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
50018           0.00              0.00              0.00             12.76   
52686           0.48             66.91             53.34             54.39   

       loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  \
50018              0.00              0.00             25.98               0.0   
52686            421.21            398.28            307.01               0.0   

       loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
50018               0.0               0.0               0.0               0.0   
52686               0.0               0.0               0.0               0.0   

       loc_og_t2c_mou_8  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
50018              0.00          0.00          0.00         38.74   
52686              2.23        488.13        451.63        361.41   

       std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  \
50018              0.00              0.00            419.03              0.00   
52686            274.19             15.06             48.49             97.53   

       std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  \
50018              0.00              7.53               0.0               0.0   
52686            154.29            186.76               0.0               0.0   

       std_og_t2f_mou_8  std_og_mou_6  std_og_mou_7  std_og_mou_8  \
50018               0.0          0.00          0.00        426.56   
52686               0.0        371.73        169.36        235.26   

       isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  spl_og_mou_6  spl_og_mou_7  \
50018           0.0           0.0           0.0           0.0           0.0   
52686           0.0           0.0           0.0           0.0           0.0   

       spl_og_mou_8  og_others_6  og_others_7  og_others_8  total_og_mou_6  \
50018          1.93          0.0          0.0          0.0            0.00   
52686          2.23          0.0          0.0          0.0          859.86   

       total_og_mou_7  total_og_mou_8  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  \
50018            0.00          467.24              0.00              0.00   
52686          620.99          598.91            133.99             71.13   

       loc_ic_t2t_mou_8  loc_ic_t2m_mou_6  loc_ic_t2m_mou_7  loc_ic_t2m_mou_8  \
50018             14.68              0.00              0.00             51.41   
52686             74.29            515.81            589.69            481.04   

       loc_ic_t2f_mou_6  loc_ic_t2f_mou_7  loc_ic_t2f_mou_8  loc_ic_mou_6  \
50018              0.00              0.00              0.00          0.00   
52686              1.56              3.03              9.01        651.38   

       loc_ic_mou_7  loc_ic_mou_8  std_ic_t2t_mou_6  std_ic_t2t_mou_7  \
50018          0.00         66.09              0.00              0.00   
52686        663.86        564.36            112.63             25.14   

       std_ic_t2t_mou_8  std_ic_t2m_mou_6  std_ic_t2m_mou_7  std_ic_t2m_mou_8  \
50018              2.59              0.00              0.00              0.00   
52686             15.04             18.39             36.66             26.96   

       std_ic_t2f_mou_6  std_ic_t2f_mou_7  std_ic_t2f_mou_8  std_ic_mou_6  

# 5 Developing new features

In [32]:
#AON: Age on network 
#Conevrt AON in years
df['aon_yr'] = round((df['aon']/365),1)

In [33]:
df.drop('aon', axis=1, inplace=True)
df.sample(2)

cust_id   arpu_6   arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  \
7965      7966  273.524  494.252  422.562        95.39       136.28   
79139    79140  355.728  235.280   58.202        77.24         6.51   

       onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
7965        142.46        225.93        417.94        492.91          11.84   
79139         6.99        272.08        225.36         35.21           0.00   

       roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  \
7965             0.0            0.0          27.91            0.0   
79139            0.0            0.0           0.00            0.0   

       roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
7965             0.0             77.31            121.44            122.63   
79139            0.0             77.24              6.51              6.99   

       loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  \
7965             183.49            387.36            431.68              0.00   
79139            270.94            225.36             35.21              1.13   

       loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
7965              17.08              3.63               0.0               0.0   
79139              0.00              0.00               0.0               0.0   

       loc_og_t2c_mou_8  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
7965                0.0        260.81        525.89        557.94   
79139               0.0        349.33        231.88         42.21   

       std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  \
7965               7.23             14.83             19.83             25.36   
79139              0.00              0.00              0.00              0.00   

       std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  \
7965              13.49             55.84               0.0               0.0   
79139              0.00              0.00               0.0               0.0   

       std_og_t2f_mou_8  std_og_mou_6  std_og_mou_7  std_og_mou_8  \
7965                0.0         32.59         28.33         75.68   
79139               0.0          0.00          0.00          0.00   

       isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  spl_og_mou_6  spl_og_mou_7  \
7965            0.0           0.0           0.0           0.0          0.00   
79139           0.0           0.0           0.0           0.0          0.36   

       spl_og_mou_8  og_others_6  og_others_7  og_others_8  total_og_mou_6  \
7965           1.75         0.00          0.0          0.0          293.41   
79139          1.34         3.71          0.0          0.0          353.04   

       total_og_mou_7  total_og_mou_8  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  \
7965           554.23          635.38             80.46            128.13   
79139          232.24           43.56             15.28             20.03   

       loc_ic_t2t_mou_8  loc_ic_t2m_mou_6  loc_ic_t2m_mou_7  loc_ic_t2m_mou_8  \
7965             134.43            183.89            280.13            286.69   
79139             10.31            156.54            413.83             81.18   

       loc_ic_t2f_mou_6  loc_ic_t2f_mou_7  loc_ic_t2f_mou_8  loc_ic_mou_6  \
7965              33.68             24.71             30.28        298.04   
79139              2.85             10.88              5.59        174.68   

       loc_ic_mou_7  loc_ic_mou_8  std_ic_t2t_mou_6  std_ic_t2t_mou_7  \
7965         432.98        451.41               0.0             14.58   
79139        444.74         97.09               0.0              0.00   

       std_ic_t2t_mou_8  std_ic_t2m_mou_6  std_ic_t2m_mou_7  std_ic_t2m_mou_8  \
7965               5.41              2.01              0.00              4.11   
79139              0.00              0.00              1.21              0.00   

       std_ic_t2f_mou_6  std_ic_t2f_mou_7  std_ic_t2f_mou_8  std_ic_mou_6  

Let's build bins for the age of network column, which will indicate the number of years a client has been utilizing the T network.

In [34]:
age_range = [ 0,  2,  4,  6,  8, 10, 12]
age_bin = [ 1, 2, 3, 4, 5, 6]
df['age_group'] = pd.cut(df['aon_yr'], age_range, labels=age_bin)
df['age_group'] = df['age_group'].astype(str)
df['age_group'].head()

0    2
1    2
2    2
3    4
4    3
Name: age_group, dtype: object

In [35]:
# let us update our categorical column list
cat_cols.append('age_group')
cat_cols

['night_pck_user_6',
 'night_pck_user_7',
 'night_pck_user_8',
 'fb_user_6',
 'fb_user_7',
 'fb_user_8',
 'age_group']

# 6 Handling Missing values

## 6.1 Checking the null values

In [36]:
null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

arpu_2g_6                   74.85
fb_user_6                   74.85
arpu_3g_6                   74.85
date_of_last_rech_data_6    74.85
night_pck_user_6            74.85
total_rech_data_6           74.85
max_rech_data_6             74.85
count_rech_2g_6             74.85
av_rech_amt_data_6          74.85
count_rech_3g_6             74.85
arpu_2g_7                   74.43
count_rech_3g_7             74.43
fb_user_7                   74.43
date_of_last_rech_data_7    74.43
max_rech_data_7             74.43
count_rech_2g_7             74.43
total_rech_data_7           74.43
arpu_3g_7                   74.43
av_rech_amt_data_7          74.43
night_pck_user_7            74.43
count_rech_3g_8             73.66
night_pck_user_8            73.66
arpu_2g_8                   73.66
fb_user_8                   73.66
date_of_last_rech_data_8    73.66
total_rech_data_8           73.66
arpu_3g_8                   73.66
av_rech_amt_data_8          73.66
max_rech_data_8             73.66
count_rech_2g_

`Observation`: After analysing I have got the 74.85% missing values for some features in this dataset.

## 6.2 Imputation of missing values

In [37]:
#Observe missing values in recharge columns 
rech_col = [i for i in df.columns if 'rech' in i]
rech_6_col = [i for i in rech_col if '6' in i]
rech_7_col = [i for i in rech_col if '7' in i]
rech_8_col = [i for i in rech_col if '8' in i]

In [38]:
#Observe missing values in recharge columns in the month of june(6)
rech_6 = pd.DataFrame(df[rech_6_col])

#adding some other columns describing data usage of customer in june(6)
vol_col = df[["vol_2g_mb_6",'vol_3g_mb_6']]

rech_6 = pd.concat([rech_6,vol_col], axis = 1) 
rech_6.sample(2)

total_rech_num_6  total_rech_amt_6  max_rech_amt_6 date_of_last_rech_6  \
16732                10               984             252          2014-06-28   
29362                 4               330             110          2014-06-25   

      date_of_last_rech_data_6  total_rech_data_6  max_rech_data_6  \
16732               2014-06-28                6.0            252.0   
29362                      NaT                NaN              NaN   

       count_rech_2g_6  count_rech_3g_6  av_rech_amt_data_6  vol_2g_mb_6  \
16732              3.0              3.0               731.0          0.0   
29362              NaN              NaN                 NaN          0.0   

       vol_3g_mb_6  
16732      2143.47  
29362         0.00

`Observation`: The missing values for max rech, count rech 2g, and 3g are displayed in the table above whenever the date of the last recharge is absent. As observed in the preceding table, the mobile internet use (2G and 3G data) for these values is zero. As there were no client recharges, I would thus impute these missing values with a value of zero.

In [39]:
#Similarly, observe missing values in recharge columns for the month of july(7)
rech_7_col

['total_rech_num_7',
 'total_rech_amt_7',
 'max_rech_amt_7',
 'date_of_last_rech_7',
 'date_of_last_rech_data_7',
 'total_rech_data_7',
 'max_rech_data_7',
 'count_rech_2g_7',
 'count_rech_3g_7',
 'av_rech_amt_data_7']

In [40]:
rech_7 = pd.DataFrame(df[rech_7_col])

#adding some other columns describing data usage of customer in july
vol_col = df[["vol_2g_mb_7",'vol_3g_mb_7']]

rech_7 = pd.concat([rech_7,vol_col], axis = 1) 
rech_7.sample(2)

total_rech_num_7  total_rech_amt_7  max_rech_amt_7 date_of_last_rech_7  \
77702                 7               462             250          2014-07-31   
14632                 9               140              50          2014-07-29   

      date_of_last_rech_data_7  total_rech_data_7  max_rech_data_7  \
77702               2014-07-31                2.0            145.0   
14632                      NaT                NaN              NaN   

       count_rech_2g_7  count_rech_3g_7  av_rech_amt_data_7  vol_2g_mb_7  \
77702              0.0              2.0               290.0        50.02   
14632              NaN              NaN                 NaN         0.00   

       vol_3g_mb_7  
77702       180.82  
14632         0.00

`Observation`: When the date of the previous recharge is absent, the associated values for max rech, count rech 2g, and 3g are also missing, as shown in the table above. According to the preceding table, the mobile internet consumption (2G and 3G data) corresponding to these numbers is zero. Therefore, I will replace these missing numbers with zero, as the consumer did not do any recharges. This resembles the trend observed throughout the month of June.

In [41]:
#Similarly,observe missing values in recharge columns in the month of August(8)
rech_8_col

['total_rech_num_8',
 'total_rech_amt_8',
 'max_rech_amt_8',
 'date_of_last_rech_8',
 'date_of_last_rech_data_8',
 'total_rech_data_8',
 'max_rech_data_8',
 'count_rech_2g_8',
 'count_rech_3g_8',
 'av_rech_amt_data_8']

In [42]:
rech_8 = pd.DataFrame(df[rech_8_col])

# adding some other columns describing data usage of customer in August
vol_col = df[["vol_2g_mb_8",'vol_3g_mb_8']]

rech_8 = pd.concat([rech_8,vol_col], axis = 1) 
rech_8.sample(2)

total_rech_num_8  total_rech_amt_8  max_rech_amt_8 date_of_last_rech_8  \
31352                 5                55              30          2014-08-28   
87300                 9               191             150          2014-08-30   

      date_of_last_rech_data_8  total_rech_data_8  max_rech_data_8  \
31352               2014-08-28                2.0             25.0   
87300               2014-08-23                5.0             52.0   

       count_rech_2g_8  count_rech_3g_8  av_rech_amt_data_8  vol_2g_mb_8  \
31352              2.0              0.0                50.0         4.96   
87300              0.0              5.0               260.0         8.93   

       vol_3g_mb_8  
31352       116.02  
87300      5148.32

`Observation`: Similar to the months of June and July, the values for max rech, count rech 2g, and 3g are missing if the date of the last recharge is absent. According to the preceding table, the mobile internet consumption (2G and 3G data) corresponding to these numbers is zero. Therefore, let's impute these missing numbers with zero, given that there were no client recharges. This is identical to the trend observed throughout the months of June and July.

## 6.3 Filling the missing value by zero: Numerical features

In [43]:
# Let's use zero for the missing recharge column data.

impute_0 = [ 'date_of_last_rech_data_6','max_rech_data_6','count_rech_2g_6','count_rech_3g_6',
           'date_of_last_rech_data_7','max_rech_data_7','count_rech_2g_7','count_rech_3g_7',
           'date_of_last_rech_data_8','max_rech_data_8','count_rech_2g_8','count_rech_3g_8']

df[impute_0] = df[impute_0].apply(lambda x: x.fillna(0))

In [44]:
# Now Checking Null values
null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

fb_user_6              74.85
arpu_2g_6              74.85
total_rech_data_6      74.85
av_rech_amt_data_6     74.85
arpu_3g_6              74.85
night_pck_user_6       74.85
fb_user_7              74.43
total_rech_data_7      74.43
arpu_2g_7              74.43
night_pck_user_7       74.43
arpu_3g_7              74.43
av_rech_amt_data_7     74.43
arpu_3g_8              73.66
arpu_2g_8              73.66
night_pck_user_8       73.66
fb_user_8              73.66
total_rech_data_8      73.66
av_rech_amt_data_8     73.66
std_og_t2m_mou_8        5.38
std_og_mou_8            5.38
std_og_t2f_mou_8        5.38
std_ic_t2m_mou_8        5.38
std_ic_mou_8            5.38
std_ic_t2f_mou_8        5.38
loc_ic_t2f_mou_8        5.38
std_ic_t2t_mou_8        5.38
isd_og_mou_8            5.38
std_og_t2t_mou_8        5.38
loc_ic_mou_8            5.38
spl_og_mou_8            5.38
og_others_8             5.38
loc_ic_t2m_mou_8        5.38
spl_ic_mou_8            5.38
loc_og_t2c_mou_8        5.38
loc_og_t2m_mou

## 6.4 Imputing Categorical columns

In [45]:
#Impute the missing values in categorical columns by (-1). 
df[cat_cols] = df[cat_cols].apply(lambda x: x.fillna(-1)) 
df[cat_cols] = df[cat_cols].astype('str')
df[cat_cols].sample(2)

night_pck_user_6 night_pck_user_7 night_pck_user_8 fb_user_6 fb_user_7  \
61726              0.0              0.0              0.0       1.0       1.0   
53569             -1.0             -1.0             -1.0      -1.0      -1.0   

      fb_user_8 age_group  
61726       1.0         2  
53569      -1.0         3

In [46]:
# Now Checking Null values
null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

total_rech_data_6      74.85
arpu_2g_6              74.85
av_rech_amt_data_6     74.85
arpu_3g_6              74.85
total_rech_data_7      74.43
av_rech_amt_data_7     74.43
arpu_3g_7              74.43
arpu_2g_7              74.43
total_rech_data_8      73.66
arpu_2g_8              73.66
arpu_3g_8              73.66
av_rech_amt_data_8     73.66
loc_ic_mou_8            5.38
std_og_t2t_mou_8        5.38
std_ic_t2t_mou_8        5.38
loc_og_mou_8            5.38
std_og_t2m_mou_8        5.38
std_ic_t2m_mou_8        5.38
loc_ic_t2f_mou_8        5.38
std_og_t2f_mou_8        5.38
std_og_mou_8            5.38
loc_og_t2c_mou_8        5.38
loc_ic_t2m_mou_8        5.38
isd_og_mou_8            5.38
loc_ic_t2t_mou_8        5.38
std_ic_t2f_mou_8        5.38
loc_og_t2f_mou_8        5.38
std_ic_mou_8            5.38
loc_og_t2m_mou_8        5.38
onnet_mou_8             5.38
offnet_mou_8            5.38
roam_ic_mou_8           5.38
roam_og_mou_8           5.38
loc_og_t2t_mou_8        5.38
ic_others_8   

## 6.5 Dropping columns with more than 40% missing values

More than forty percent of the average revenue per user during both the good and action phases are missing. Since the focus of this research is the identification of consumers who are likely to churn based only on usage-based churn, I may safely exclude these columns without compromising the research.

In [47]:
miss = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False) 

miss = pd.DataFrame(miss[miss >= 40])
threshold_col = miss.index

df = df.drop(threshold_col,1)

In [48]:
df.shape

(99999, 152)

## 6.6 Dropping missing values for out going calls for month of June, July, and August

#### For Month of June

In [49]:
#According to my business domain knowledge, the total_og column represents the sum of local, standard, special, and other outgoing calls.
og_call_6 = ['loc_og_mou_6','std_og_mou_6','isd_og_mou_6','spl_og_mou_6','og_others_6','total_og_mou_6']
total_og_6 = df[og_call_6]

#limiting the filtering to clients who have made no outbound calls.
total_og_6.loc[total_og_6['total_og_mou_6']==0].sample(2)

loc_og_mou_6  std_og_mou_6  isd_og_mou_6  spl_og_mou_6  og_others_6  \
89766           0.0           0.0           0.0           0.0          0.0   
74278           0.0           0.0           0.0           0.0          0.0   

       total_og_mou_6  
89766             0.0  
74278             0.0

`Observation`: Based on my understanding of the business area and the preceding table, I can deduce that when the total number of outgoing calls for a given month is 0, no outgoing calls have been made. I am going to change the missing values for local outgoing calls, normal outgoing calls, special outgoing calls, and other outgoing calls with 0.

In [50]:
#Let's see whether our assumption is correct or not: the sum of all outgoing calls equals the 'total_og_mou_6'.

df['outgoing_total_6'] = df['loc_og_mou_6']+ df['std_og_mou_6']+df['isd_og_mou_6']+df['spl_og_mou_6']+df['og_others_6']
df[['outgoing_total_6','total_og_mou_6']].dropna().corr()

outgoing_total_6  total_og_mou_6
outgoing_total_6               1.0             1.0
total_og_mou_6                 1.0             1.0

`Observation`: From the above table, it is evident that these variables are associated; thus, when the total number of outgoing calls for a given month is 0,that means no outbound calls have been made. as a result, I am going to replace the missing values for 'local outgoing calls', 'normal outgoing calls', 'special outgoing calls', and 'other outgoing calls' with 0.

In [51]:
#Imputing by '0'

df[og_call_6] = df[og_call_6].fillna(0)

### local outgoing calls(T2T,T2M,T2F & T2C)

In [52]:
og_loc_6 = ['loc_og_t2t_mou_6','loc_og_t2m_mou_6','loc_og_t2f_mou_6','loc_og_t2c_mou_6','loc_og_mou_6']
loc_og_6 = df[og_loc_6]

# filtering only those clients who didn't make any 'local outgoing calls'
loc_og_6.loc[loc_og_6 ['loc_og_mou_6']==0].sample(2)

loc_og_t2t_mou_6  loc_og_t2m_mou_6  loc_og_t2f_mou_6  loc_og_t2c_mou_6  \
16942               0.0               0.0               0.0               0.0   
56668               NaN               NaN               NaN               NaN   

       loc_og_mou_6  
16942           0.0  
56668           0.0

Local outbound calls in any given month are equal to the total of all local outgoing call kinds, i.e. (T2T,T2M,T2F & T2C)

The monthly total number of local outbound calls (T2T,T2M,T2F, and T2C) is zero. Now, I will replace the missing value with zero.

In [53]:
#Lets verify is the assumption right or not, sum of 'all local outgoing calls' is 
#equal to the the 'total local outgoing call'
#"loc_og_mou_6"

df['loc_outgoing_total_6'] = df['loc_og_t2t_mou_6']+ df['loc_og_t2m_mou_6']+df['loc_og_t2f_mou_6']+df['loc_og_t2c_mou_6']
df[['loc_outgoing_total_6','loc_og_mou_6']].dropna().corr().round()

loc_outgoing_total_6  loc_og_mou_6
loc_outgoing_total_6                   1.0           1.0
loc_og_mou_6                           1.0           1.0

`Observation:` From the above table, it is evident that these variables are associated; thus, when the total number of local outgoing calls for a given month is 0,that means no local outbound calls have been made. as a result, I am going to replace the missing values for T2T,T2M,T2F & T2C by zero

In [54]:
# Imputing by '0'
df[og_loc_6] = df[og_loc_6].fillna(0)

#### total standard outgoing calls 

In [55]:
og_std_6 = ['std_og_t2t_mou_6','std_og_t2m_mou_6','std_og_t2f_mou_6','std_og_mou_6']
std_og_6 = df[og_std_6]
std_og_6.sample(5)

std_og_t2t_mou_6  std_og_t2m_mou_6  std_og_t2f_mou_6  std_og_mou_6
82107               0.0              8.91               0.0          8.91
71003               0.0              0.00               0.0          0.00
11283               0.0             44.96               0.0         44.96
71242               0.0              8.74               0.0          8.74
19744               0.0              2.50               0.0          2.50

In [56]:
# Imputing by '0'
df[og_std_6] = df[og_std_6].fillna(0)

#### For Month of July

Similar process is applied like June

#### Total Outgoing Call

In [57]:
# From the business domain knowledge, we know that total_og column is the addition of local , standard, special and other outgoing calls
og_call_7 = ['loc_og_mou_7','std_og_mou_7','isd_og_mou_7','spl_og_mou_7','og_others_7','total_og_mou_7']
total_og_7 = df[og_call_7]
# filtering only those clients who have made no outgoing calls
total_og_7.loc[total_og_7['total_og_mou_7']==0].sample(2)

loc_og_mou_7  std_og_mou_7  isd_og_mou_7  spl_og_mou_7  og_others_7  \
54864           0.0           0.0           0.0           0.0          0.0   
50682           NaN           NaN           NaN           NaN          NaN   

       total_og_mou_7  
54864             0.0  
50682             0.0

In [58]:
#Lets verify is our assumption right or not, sum of all outgoing calls is equal to the the tota_og_mou_7
df['outgoing_total_7'] = df['loc_og_mou_7']+ df['std_og_mou_7']+df['isd_og_mou_7']+df['spl_og_mou_7']+df['og_others_7']
df[['outgoing_total_7','total_og_mou_7']].dropna().corr()

outgoing_total_7  total_og_mou_7
outgoing_total_7               1.0             1.0
total_og_mou_7                 1.0             1.0

In [59]:
# Imputing by '0'
df[og_call_7] = df[og_call_7].fillna(0)

In [60]:
og_loc_7 = ['loc_og_t2t_mou_7','loc_og_t2m_mou_7','loc_og_t2f_mou_7','loc_og_t2c_mou_7','loc_og_mou_7']
loc_og_7 = df[og_loc_7]

# filtering only those clients who didn't made any local outgoing calls
loc_og_7.loc[loc_og_7 ['loc_og_mou_7']==0].sample(3)

loc_og_t2t_mou_7  loc_og_t2m_mou_7  loc_og_t2f_mou_7  loc_og_t2c_mou_7  \
19509               0.0               0.0               0.0               0.0   
37097               NaN               NaN               NaN               NaN   
40110               NaN               NaN               NaN               NaN   

       loc_og_mou_7  
19509           0.0  
37097           0.0  
40110           0.0

#### T2T,T2M,T2F & T2C

In [61]:
#Lets verify is our assumption right or not, sum of all local outgoing calls is equal to 
#the total local outgoing call
#"loc_og_mou_7"

df['loc_outgoing_total_7'] = df['loc_og_t2t_mou_7']+ df['loc_og_t2m_mou_7']+df['loc_og_t2f_mou_7']+df['loc_og_t2c_mou_7']
df[['loc_outgoing_total_7','loc_og_mou_7']].dropna().corr().round()

loc_outgoing_total_7  loc_og_mou_7
loc_outgoing_total_7                   1.0           1.0
loc_og_mou_7                           1.0           1.0

In [62]:
# Imputing by '0'
df[og_loc_7] = df[og_loc_7].fillna(0)

In [63]:
og_std_7 = ['std_og_t2t_mou_7','std_og_t2m_mou_7','std_og_t2f_mou_7','std_og_mou_7']
std_og_7 = df[og_std_7]
std_og_7.sample(3)

std_og_t2t_mou_7  std_og_t2m_mou_7  std_og_t2f_mou_7  std_og_mou_7
22466             18.58              0.00              0.00         18.58
36947              0.00             16.93              0.00         16.93
52202            106.14             25.48              5.28        136.91

#### Total standard outgoing calls 

In [64]:
#Lets verify is our assumption right or not, sum of all std outgoing calls is equal to 
#the total std outgoing call
#"std_og_mou_7"

df['std_outgoing_total_7'] = df['std_og_t2t_mou_7']+ df['std_og_t2m_mou_7']+df['std_og_t2f_mou_7']
df[['std_outgoing_total_7','std_og_mou_7']].dropna().corr()

std_outgoing_total_7  std_og_mou_7
std_outgoing_total_7                   1.0           1.0
std_og_mou_7                           1.0           1.0

In [65]:
# Imputing by '0'
df[og_std_7] = df[og_std_7].fillna(0)

#### For Month of August

Similar process is applied like June

#### Total Outgoing Call

In [66]:
# From the business domain knowledge, we know that total_og column is the addition of local , standard, special and other outgoing calls
og_call_8 = ['loc_og_mou_8','std_og_mou_8','isd_og_mou_8','spl_og_mou_8','og_others_8','total_og_mou_8']
total_og_8 = df[og_call_8]
# filtering only those clients who have made no outgoing calls
total_og_8.loc[total_og_8['total_og_mou_8']==0].sample(3)

loc_og_mou_8  std_og_mou_8  isd_og_mou_8  spl_og_mou_8  og_others_8  \
6291            0.0           0.0           0.0           0.0          0.0   
74175           NaN           NaN           NaN           NaN          NaN   
68723           0.0           0.0           0.0           0.0          0.0   

       total_og_mou_8  
6291              0.0  
74175             0.0  
68723             0.0

In [67]:
#Lets verify is our assumption right or not, sum of 
#all outgoing calls is equal to the tota_og_mou_8

df['outgoing_total_8'] = df['loc_og_mou_8']+ df['std_og_mou_8']+df['isd_og_mou_8']+df['spl_og_mou_8']+df['og_others_8']
df[['outgoing_total_8','total_og_mou_8']].dropna().corr()

outgoing_total_8  total_og_mou_8
outgoing_total_8               1.0             1.0
total_og_mou_8                 1.0             1.0

In [68]:
# Imputing by '0'
df[og_call_8] = df[og_call_8].fillna(0)

#### T2T,T2M,T2F & T2C

In [69]:
og_loc_8 = ['loc_og_t2t_mou_8','loc_og_t2m_mou_8','loc_og_t2f_mou_8','loc_og_t2c_mou_8','loc_og_mou_8']
loc_og_8 = df[og_loc_8]

# filtering only those clients who have made no local outgoing calls
loc_og_8.loc[loc_og_8 ['loc_og_mou_8']==0].sample(2)

loc_og_t2t_mou_8  loc_og_t2m_mou_8  loc_og_t2f_mou_8  loc_og_t2c_mou_8  \
6743                NaN               NaN               NaN               NaN   
77326               NaN               NaN               NaN               NaN   

       loc_og_mou_8  
6743            0.0  
77326           0.0

In [70]:
#Lets verify is our assumption right or not, sum of all local outgoing calls is equal to the the total local outgoing call
#"loc_og_mou_8"

df['loc_outgoing_total_8'] = df['loc_og_t2t_mou_8']+ df['loc_og_t2m_mou_8']+df['loc_og_t2f_mou_8']+df['loc_og_t2c_mou_8']
df[['loc_outgoing_total_8','loc_og_mou_8']].dropna().corr().round()

loc_outgoing_total_8  loc_og_mou_8
loc_outgoing_total_8                   1.0           1.0
loc_og_mou_8                           1.0           1.0

In [71]:
# Imputing by '0'
df[og_loc_8] = df[og_loc_8].fillna(0)

#### std outgoing calls

In [72]:
og_std_8 = ['std_og_t2t_mou_8','std_og_t2m_mou_8','std_og_t2f_mou_8','std_og_mou_8']
std_og_8 = df[og_std_8]
std_og_8.sample(3)

std_og_t2t_mou_8  std_og_t2m_mou_8  std_og_t2f_mou_8  std_og_mou_8
82923               NaN               NaN               NaN           0.0
2613                0.0               0.0               0.0           0.0
68434               0.0               0.0               0.0           0.0

In [73]:
#Lets verify is our assumption right or not, sum of all std outgoing 
#calls is equal to the total std outgoing call
#"std_og_mou_8"

df['std_outgoing_total_8'] = df['std_og_t2t_mou_8']+ df['std_og_t2m_mou_8']+df['std_og_t2f_mou_8']
df[['std_outgoing_total_8','std_og_mou_8']].dropna().corr()

std_outgoing_total_8  std_og_mou_8
std_outgoing_total_8                   1.0           1.0
std_og_mou_8                           1.0           1.0

In [74]:
# Imputing by '0'
df[og_std_8] = df[og_std_8].fillna(0)

## 6.7 Dropping missing values for incoming calls for month of June, July, and August

usage for local, standard, special and others¶

#### For month of June

In [75]:
# From the business domain knowledge, we know that total_ic column is the addition of local , standard, special andd other incoming calls
ic_call_6 = ['loc_ic_mou_6','std_ic_mou_6','isd_ic_mou_6','spl_ic_mou_6','ic_others_6','total_ic_mou_6']
total_ic_6 = df[ic_call_6]
# filtering only those clients who have made no outgoing calls
total_ic_6 .loc[total_ic_6 ['total_ic_mou_6']==0].sample(2)

loc_ic_mou_6  std_ic_mou_6  isd_ic_mou_6  spl_ic_mou_6  ic_others_6  \
23294           NaN           NaN           NaN           NaN          NaN   
67668           NaN           NaN           NaN           NaN          NaN   

       total_ic_mou_6  
23294             0.0  
67668             0.0

`Observation`: Based on my understanding of the business domain and the aforementioned statistics, I can deduce that when the total number of incoming calls is zero, no incoming calls are received during that month. The missing values can be replaced with 0 in the columns for local incoming calls, regular incoming calls, special incoming calls, and other incoming calls.

In [76]:
#Lets verify is our assumption right or not, sum of all incoming calls is equal to the tota_ic_mou_6

df['incoming_total_6'] = df['loc_ic_mou_6']+ df['std_ic_mou_6']+df['isd_ic_mou_6']+df['spl_ic_mou_6']+df['ic_others_6']
df[['incoming_total_6','total_ic_mou_6']].dropna().corr()

incoming_total_6  total_ic_mou_6
incoming_total_6               1.0             1.0
total_ic_mou_6                 1.0             1.0

`Observation`: As can be seen in the table above, when the total number of incoming calls is 0, no outgoing calls have been made during that month. 0 can be substituted for missing values in the columns for local outgoing calls, normal outgoing calls, special incoming calls, and other outgoing calls.

In [77]:
# Imputing by '0'
df[ic_call_6] = df[ic_call_6].fillna(0)

#### T2T,T2M,T2F & T2C

In [78]:
ic_loc_6 = ['loc_ic_t2t_mou_6','loc_ic_t2m_mou_6','loc_ic_t2f_mou_6','loc_ic_mou_6']
loc_ic_6= df[ic_loc_6]

# filtering only those clients who have made no local outgoing calls
loc_ic_6.loc[loc_ic_6['loc_ic_mou_6']==0].sample(3)

loc_ic_t2t_mou_6  loc_ic_t2m_mou_6  loc_ic_t2f_mou_6  loc_ic_mou_6
82601               NaN               NaN               NaN           0.0
33061               NaN               NaN               NaN           0.0
10655               NaN               NaN               NaN           0.0

`Observation`:If the total number of local calls received in a particular month is zero, no local calls were received, i.e. T2T, T2M, T2F, and T2C. I am able to replace '0' for the missing equivalent numbers.

In [79]:
#Lets verify is our assumption right or not, sum of all local outgoing calls is equal to the 
#total local outgoing call
#"loc_ic_mou_6"

df['loc_incoming_total_6'] = df['loc_ic_t2t_mou_6']+ df['loc_ic_t2m_mou_6']+df['loc_ic_t2f_mou_6']
df[['loc_incoming_total_6','loc_ic_mou_6']].dropna().corr()

loc_incoming_total_6  loc_ic_mou_6
loc_incoming_total_6                   1.0           1.0
loc_ic_mou_6                           1.0           1.0

In [80]:
# Imputing by '0'
df[ic_loc_6] = df[ic_loc_6].fillna(0)

#### std incoming calls

It is obvious from the statistics that standard incoming calls in any given month equal the total of all standard incoming call categories, i.e. (T2T,T2M,T2F & T2C)

In [81]:
ic_std_6 = ['std_ic_t2t_mou_6','std_ic_t2m_mou_6','std_ic_t2f_mou_6','std_ic_mou_6']
std_ic_6 = df[ic_std_6]
std_ic_6.sample(3)

std_ic_t2t_mou_6  std_ic_t2m_mou_6  std_ic_t2f_mou_6  std_ic_mou_6
73272               0.0              0.15               0.0          0.15
65806               0.0              0.00               0.0          0.00
46858               NaN               NaN               NaN          0.00

`Observation`:If the total number of regular outbound calls in a particular month is zero, no T2T, T2M, T2F, or T2C calls were made. I am able to substitute '0' for the missing values that correspond to them.

In [82]:
#Lets verify is our assumption right or not, sum of all std outgoing calls is equal to the total std outgoing call
#"std_ic_mou_6"

df['std_incoming_total_6'] = df['std_ic_t2t_mou_6']+ df['std_ic_t2m_mou_6']+df['std_ic_t2f_mou_6']
df[['std_incoming_total_6','std_ic_mou_6']].dropna().corr()

std_incoming_total_6  std_ic_mou_6
std_incoming_total_6                   1.0           1.0
std_ic_mou_6                           1.0           1.0

`Observation`:If the total number of standard outbound calls in a particular month is 0, then no standard calls, i.e. T2T, T2M, T2F, and T2C, were made.

In [83]:
# Imputing by '0'
df[ic_std_6] = df[ic_std_6].fillna(0)

#### For month of July

# Total Calls

In [84]:
# From the business domain knowledge, we know that total_ic column is the addition of local , standard, special andd other incoming calls
ic_call_7 = ['loc_ic_mou_7','std_ic_mou_7','isd_ic_mou_7','spl_ic_mou_7','ic_others_7','total_ic_mou_7']
total_ic_7 = df[ic_call_7]
# filtering only those clients who have made no outgoing calls
total_ic_7 .loc[total_ic_7 ['total_ic_mou_7']==0].sample(2)

loc_ic_mou_7  std_ic_mou_7  isd_ic_mou_7  spl_ic_mou_7  ic_others_7  \
19983           0.0           0.0           0.0           0.0          0.0   
91902           NaN           NaN           NaN           NaN          NaN   

       total_ic_mou_7  
19983             0.0  
91902             0.0

In [85]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the the total local incoming call
#"loc_ic_mou_7"

df['loc_incoming_total_7'] = df['loc_ic_mou_7']+ df['std_ic_mou_7']+df['isd_ic_mou_7']+df['spl_ic_mou_7']+df['ic_others_7']
df[['loc_incoming_total_7','total_ic_mou_7']].dropna().corr()

loc_incoming_total_7  total_ic_mou_7
loc_incoming_total_7                   1.0             1.0
total_ic_mou_7                         1.0             1.0

In [86]:
# Imputing by '0'
df[ic_call_7] = df[ic_call_7].fillna(0)

#### T2T,T2M,T2F & T2C

In [87]:
ic_loc_7 = ['loc_ic_t2t_mou_7','loc_ic_t2m_mou_7','loc_ic_t2f_mou_7','loc_ic_mou_7']
loc_ic_7= df[ic_loc_7]

# filtering only those clients who have made no local outgoing calls
loc_ic_7.loc[loc_ic_7['loc_ic_mou_7']==0].sample(3)

loc_ic_t2t_mou_7  loc_ic_t2m_mou_7  loc_ic_t2f_mou_7  loc_ic_mou_7
3394                NaN               NaN               NaN           0.0
80316               0.0               0.0               0.0           0.0
27619               0.0               0.0               0.0           0.0

In [88]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the 
#total local incoming call
#"loc_ic_mou_7"

df['loc_incoming_total_7'] = df['loc_ic_t2t_mou_7']+ df['loc_ic_t2m_mou_7']+df['loc_ic_t2f_mou_7']
df[['loc_incoming_total_7','loc_ic_mou_7']].dropna().corr()

loc_incoming_total_7  loc_ic_mou_7
loc_incoming_total_7                   1.0           1.0
loc_ic_mou_7                           1.0           1.0

In [89]:
# Imputing by '0'
df[ic_loc_7] = df[ic_loc_7].fillna(0)

#### Standard Outgoing calls

In [90]:
ic_std_7 = ['std_ic_t2t_mou_7','std_ic_t2m_mou_7','std_ic_t2f_mou_7','std_ic_mou_7']
std_ic_7 = df[ic_std_7]
std_ic_7.sample(2)

std_ic_t2t_mou_7  std_ic_t2m_mou_7  std_ic_t2f_mou_7  std_ic_mou_7
84305              0.00              1.83               0.0          1.83
69718              6.61             66.99               0.0         73.61

In [91]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the the total local incoming call
#"std_ic_mou_7"

df['std_incoming_total_7'] = df['std_ic_t2t_mou_7']+ df['std_ic_t2m_mou_7']+df['std_ic_t2f_mou_7']
df[['std_incoming_total_7','std_ic_mou_7']].dropna().corr()

std_incoming_total_7  std_ic_mou_7
std_incoming_total_7                   1.0           1.0
std_ic_mou_7                           1.0           1.0

In [92]:
# Imputing by '0'
df[ic_std_7] = df[ic_std_7].fillna(0)

#### For August Month

In [93]:
# From the business domain knowledge, we know that total_ic column is the addition of local , standard, special andd other incoming calls
ic_call_8 = ['loc_ic_mou_8','std_ic_mou_8','isd_ic_mou_8','spl_ic_mou_8','ic_others_8','total_ic_mou_8']
total_ic_8 = df[ic_call_8]
# filtering only those clients who have made no outgoing calls
total_ic_8 .loc[total_ic_8['total_ic_mou_8']==0].sample(3)

loc_ic_mou_8  std_ic_mou_8  isd_ic_mou_8  spl_ic_mou_8  ic_others_8  \
32010           NaN           NaN           NaN           NaN          NaN   
70267           0.0           0.0           0.0           0.0          0.0   
75919           NaN           NaN           NaN           NaN          NaN   

       total_ic_mou_8  
32010             0.0  
70267             0.0  
75919             0.0

In [94]:
#Lets verify is our assumption right or not, sum of all incoming calls is equal to the the tota_ic_mou_8

df['incoming_total_8'] = df['loc_ic_mou_8']+ df['std_ic_mou_8']+df['isd_ic_mou_8']+df['spl_ic_mou_8']+df['ic_others_8']
df[['incoming_total_8','total_ic_mou_8']].dropna().corr()

incoming_total_8  total_ic_mou_8
incoming_total_8               1.0             1.0
total_ic_mou_8                 1.0             1.0

In [95]:
# Imputing by '0'
df[ic_call_8] = df[ic_call_8].fillna(0)

In [96]:
ic_loc_8 = ['loc_ic_t2t_mou_8','loc_ic_t2m_mou_8','loc_ic_t2f_mou_8','loc_ic_mou_8']
loc_ic_8= df[ic_loc_8]

# filtering only those clients who have made no local outgoing calls
loc_ic_8.loc[loc_ic_8['loc_ic_mou_8']==0].sample(3)

loc_ic_t2t_mou_8  loc_ic_t2m_mou_8  loc_ic_t2f_mou_8  loc_ic_mou_8
69324               0.0               0.0               0.0           0.0
68722               0.0               0.0               0.0           0.0
21345               NaN               NaN               NaN           0.0

In [97]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the the total local incoming call
#"loc_ic_mou_8"

df['loc_incoming_total_8'] = df['loc_ic_t2t_mou_8']+ df['loc_ic_t2m_mou_8']+df['loc_ic_t2f_mou_8']
df[['loc_incoming_total_8','loc_ic_mou_8']].dropna().corr()

loc_incoming_total_8  loc_ic_mou_8
loc_incoming_total_8                   1.0           1.0
loc_ic_mou_8                           1.0           1.0

In [98]:
# Imputing by '0'
df[ic_loc_8] = df[ic_loc_8].fillna(0)

In [99]:
ic_std_8 = ['std_ic_t2t_mou_8','std_ic_t2m_mou_8','std_ic_t2f_mou_8','std_ic_mou_8']
std_ic_8 = df[ic_std_8]
std_ic_8.sample(3)

std_ic_t2t_mou_8  std_ic_t2m_mou_8  std_ic_t2f_mou_8  std_ic_mou_8
6989              11.99           1446.88              0.99       1459.88
41725             21.84             48.46              0.00         70.31
48923              0.00              3.93              0.00          3.93

In [100]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the the total local incoming call
#"std_ic_mou_8"

df['std_incoming_total_8'] = df['std_ic_t2t_mou_8']+ df['std_ic_t2m_mou_8']+df['std_ic_t2f_mou_8']
df[['std_incoming_total_8','std_ic_mou_8']].dropna().corr()

std_incoming_total_8  std_ic_mou_8
std_incoming_total_8                   1.0           1.0
std_ic_mou_8                           1.0           1.0

In [101]:
# Imputing by '0'
df[ic_std_8] = df[ic_std_8].fillna(0)

#### Check the Null Values

In [102]:
## Let us now check the % of null values
null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

std_incoming_total_8    5.38
offnet_mou_8            5.38
outgoing_total_8        5.38
std_outgoing_total_8    5.38
roam_og_mou_8           5.38
roam_ic_mou_8           5.38
loc_outgoing_total_8    5.38
loc_incoming_total_8    5.38
onnet_mou_8             5.38
incoming_total_8        5.38
offnet_mou_6            3.94
roam_ic_mou_6           3.94
outgoing_total_6        3.94
loc_outgoing_total_6    3.94
roam_og_mou_6           3.94
onnet_mou_6             3.94
std_incoming_total_6    3.94
loc_incoming_total_6    3.94
incoming_total_6        3.94
outgoing_total_7        3.86
offnet_mou_7            3.86
std_outgoing_total_7    3.86
loc_outgoing_total_7    3.86
std_incoming_total_7    3.86
roam_og_mou_7           3.86
onnet_mou_7             3.86
roam_ic_mou_7           3.86
loc_incoming_total_7    3.86
date_of_last_rech_8     3.62
date_of_last_rech_7     1.77
date_of_last_rech_6     1.61
dtype: float64

## 3.9 Drop Null Values from Rows

In [103]:
df = df.dropna(how='all',axis=0) 

In [104]:
df.shape

(99999, 168)

In [105]:
# Now Checking Null values
null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

std_incoming_total_8    5.38
offnet_mou_8            5.38
outgoing_total_8        5.38
std_outgoing_total_8    5.38
roam_og_mou_8           5.38
roam_ic_mou_8           5.38
loc_outgoing_total_8    5.38
loc_incoming_total_8    5.38
onnet_mou_8             5.38
incoming_total_8        5.38
offnet_mou_6            3.94
roam_ic_mou_6           3.94
outgoing_total_6        3.94
loc_outgoing_total_6    3.94
roam_og_mou_6           3.94
onnet_mou_6             3.94
std_incoming_total_6    3.94
loc_incoming_total_6    3.94
incoming_total_6        3.94
outgoing_total_7        3.86
offnet_mou_7            3.86
std_outgoing_total_7    3.86
loc_outgoing_total_7    3.86
std_incoming_total_7    3.86
roam_og_mou_7           3.86
onnet_mou_7             3.86
roam_ic_mou_7           3.86
loc_incoming_total_7    3.86
date_of_last_rech_8     3.62
date_of_last_rech_7     1.77
date_of_last_rech_6     1.61
dtype: float64

## 3.10 Drop Null Values from features¶

In [106]:
# Let us drop the date columns as they do not infer anything 
del_date = [i for i in df.columns if 'date' in i]
df= df.drop(del_date,1)


In [107]:
df.shape

(99999, 162)

In [108]:
# Now Checking Null values
null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

std_incoming_total_8    5.38
roam_og_mou_8           5.38
loc_incoming_total_8    5.38
incoming_total_8        5.38
outgoing_total_8        5.38
onnet_mou_8             5.38
loc_outgoing_total_8    5.38
std_outgoing_total_8    5.38
offnet_mou_8            5.38
roam_ic_mou_8           5.38
loc_outgoing_total_6    3.94
incoming_total_6        3.94
loc_incoming_total_6    3.94
std_incoming_total_6    3.94
outgoing_total_6        3.94
roam_og_mou_6           3.94
onnet_mou_6             3.94
roam_ic_mou_6           3.94
offnet_mou_6            3.94
roam_og_mou_7           3.86
loc_outgoing_total_7    3.86
std_incoming_total_7    3.86
outgoing_total_7        3.86
std_outgoing_total_7    3.86
onnet_mou_7             3.86
offnet_mou_7            3.86
roam_ic_mou_7           3.86
loc_incoming_total_7    3.86
dtype: float64

### Replacing the remaining missing values¶

Replace NaN values with the column's mean if the missing data percentage is less than 4%. Because the great majority of the data is valid, replacing the remaining values with the mean will have minimal impact on the research.

In [109]:
df[null.index].describe()

std_incoming_total_8  roam_og_mou_8  loc_incoming_total_8  \
count          94621.000000   94621.000000          94621.000000   
mean              33.152136       9.971890            167.423711   
std              110.125911      64.713221            250.023926   
min                0.000000       0.000000              0.000000   
25%                0.010000       0.000000             32.740000   
50%                5.880000       0.000000             93.810000   
75%               27.690000       0.000000            207.270000   
max             5957.130000    5337.040000          10830.160000   

       incoming_total_8  outgoing_total_8   onnet_mou_8  loc_outgoing_total_8  \
count      94621.000000      94621.000000  94621.000000          94621.000000   
mean         210.040473        321.398096    133.018098            142.754745   
std          293.412497        485.864968    308.951589            246.156285   
min            0.000000          0.000000      0.000000              0.000000   
25%           49.110000         51.580000      6.460000             18.620000   
50%          124.930000        153.790000     32.360000             65.180000   
75%          260.660000        391.960000    115.860000            167.650000   
max        10830.370000      14043.050000  10752.560000          11040.560000   

       std_outgoing_total_8  offnet_mou_8  roam_ic_mou_8  \
count          94621.000000  94621.000000   94621.000000   
mean             174.188855    196.574803       7.292981   
std              411.631534    327.170662      68.402466   
min                0.000000      0.000000       0.000000   
25%                0.000000     31.630000       0.000000   
50%               10.410000     92.140000       0.000000   
75%              147.940000    228.260000       0.000000   
max            13980.050000  14007.340000   13095.360000   

       loc_outgoing_total_6  incoming_total_6  loc_incoming_total_6  \
count          96062.000000      96062.000000          96062.000000   
mean             145.316920        208.325058            167.482320   
std              252.007954        294.681705            254.122393   
min                0.000000          0.000000              0.000000   
25%               18.300000         46.680000             30.382500   
50%               66.230000        122.240000             92.145000   
75%              169.417500        259.840000            208.065000   
max            10645.270000       7716.130000           7454.620000   

       std_incoming_total_6  outgoing_total_6  roam_og_mou_6   onnet_mou_6  \
count          96062.000000      96062.000000   96062.000000  96062.000000   
mean              32.454631        317.631638      13.911337    132.395875   
std              106.282273        468.599116      71.443196    297.207406   
min                0.000000          0.000000       0.000000      0.000000   
25%                0.000000         54.360000       0.000000      7.380000   
50%                5.890000        156.320000       0.000000     34.310000   
75%               26.930000        388.590000       0.000000    118.740000   
max             5712.100000      10674.020000    3775.110000   7376.710000   

       roam_ic_mou_6  offnet_mou_6  roam_og_mou_7  loc_outgoing_total_7  \
count   96062.000000  96062.000000   96140.000000          96140.000000   
mean        9.950013    197.935577       9.818732            143.031627   
std        72.825411    316.851613      58.455762            248.984326   
min         0.000000      0.000000       0.000000              0.000000   
25%         0.000000     34.730000       0.000000             18.930000   
50%         0.000000     96.310000       0.000000             65.110000   
75%         0.000000    231.860000       0.000000            166.090000   
max     13724.380000   8362.360000    2812.040000           7674.760000   

       std_incoming_total_7  outgoing_total_7  std_outgoing_total_7  \
count          96140.000000      

In [110]:
miss_col = null.index
for i in miss_col:
    df[i] = df[i].fillna(df[i].mean())

In [111]:
## Let us now check the % of null values

null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

Series([], dtype: float64)

`Observation`: There is no missing values in this dataset